Added support of multi-page highlights

jonathanbyrne opened the pull request:
Also, added a launch config to debug inside VSCode because it isn't very obvious how to do that.

agentcooper commented on src/types.ts:
```diff
@@ -16,11 +20,13 @@ export interface Scaled {
-  rects: Array<LTWH>;
+  boundingRect: LTWHP;
+  rects: Array<LTWHP>;
   pageNumber: number;
```
Do we still need `pageNumber` here if all the rects have `pageNumber`?

jonathanbyrne replied:
I believe you are correct.  I will check that out.

jonathanbyrne replied:
I looked at this further and my thought is that `pageNumber` should probably stay for the purpose of backwards compatibility.  If it is removed, older versions of the highlights would not display correctly.  We store the highlights in a database and display them when a PDF is loaded.  I suspect a lot of other people do the same.

agentcooper commented:
Hey, this is pretty good!

I did some testing, in all cases I've highlighted from the end of page 3 `Fast type checking. The key to Flow’s speed is modularity...` till the beginning of page 4 `...good software engineering practice anyway.`

1. It seems that the scroll to position does not work right for multi-page highlights. With the highlight mentioned above, when I click the sidebar note, the browser jumps to the beginning of page 3, but the highlight is not visible. I can reproduce this in both Brave (Chrome based) and Safari.

2. In Safari the whole next page is highlighted. I wonder if there is some quick fix that can be applied to mitigate that?
<img width="1552" alt="image" src="https://user-images.githubusercontent.com/794591/139811467-df7c8073-89c3-4e9e-a4d8-2ce0ff892fda.png">

jonathanbyrne commented:
> Hey, this is pretty good!

Thank you!

> 
> I did some testing, in all cases I've highlighted from the end of page 3 `Fast type checking. The key to Flow’s speed is modularity...` till the beginning of page 4 `...good software engineering practice anyway.`
> 
> 1. It seems that the scroll to position does not work right for multi-page highlights. With the highlight mentioned above, when I click the sidebar note, the browser jumps to the beginning of page 3, but the highlight is not visible. I can reproduce this in both Brave (Chrome based) and Safari.

I will get that fixed.

> 2. In Safari the whole next page is highlighted. I wonder if there is some quick fix that can be applied to mitigate that?
> 
> <img alt="image" width="1552" src="https://user-images.githubusercontent.com/794591/139811467-df7c8073-89c3-4e9e-a4d8-2ce0ff892fda.png">

I will fix this too.  I forgot all about testing in Safari.

> It would also be great if you can add a test about the multi-page highlight to the `./tests/integration.test.js`.

🤦‍♂️  I am sorry about that.  I had every intention of writing a test for my change but completely forgot about it at the end.

jonathanbyrne commented:
> > 1. It seems that the scroll to position does not work right for multi-page highlights. With the highlight mentioned above, when I click the sidebar note, the browser jumps to the beginning of page 3, but the highlight is not visible. I can reproduce this in both Brave (Chrome based) and Safari.
> 
> I will get that fixed.

This is now fixed.

> > 1. In Safari the whole next page is highlighted. I wonder if there is some quick fix that can be applied to mitigate that?
> 
> I will fix this too. I forgot all about testing in Safari.
> 

I am 95% sure I found the problem and fixed it.  However, I do not have a Mac to test on, so I cannot be 100% certain.

> > It would also be great if you can add a test about the multi-page highlight to the `./tests/integration.test.js`.
> 
> 🤦‍♂️ I am sorry about that. I had every intention of writing a test for my change but completely forgot about it at the end.

It's not the greatest test in the world, but I think it confirms that multipage selects work.  This was my first time using Puppeteer and I couldn't debug the test while the browser was visible so I was flying blind.  I'm running Ubuntu on WSL so things can get a bit goofy.

jonathanbyrne commented:
Sorry about those build errors.  It should be all fixed now.

agentcooper commented:
Great work, thanks! Released in 5.3.0.
